File: lab4/src/Polynomial.java
import java.util.ArrayList;
import java.util.List;

public class Polynomial {
    List<Integer> coefficients;
    public Polynomial(List<Integer> c) {
        this.coefficients = c;
    }
    public Polynomial(int degree) {
        this.coefficients = new ArrayList<>();
        for( int i=0; i < degree; i ++){
            coefficients.add(0);
        }
    }
    public int getDegree(){
        return this.coefficients.size() - 1;
    }
    public List<Integer> getCoefficients(){
        return coefficients;
    }
    public void setCoefficient(int index, int value){
        this.coefficients.set(index, value);
    }
    @Override
    public String toString() {
        return "Polynomial{" +
                "coefficients=" + coefficients +
                '}';
    }

    public Polynomial minus(Polynomial a, Polynomial b) {
        int actualsize = a.getDegree();
        int maxlen = Math.max(b.getDegree(), a.getDegree());
        List<Integer> newCoefficients = new ArrayList<>();
        for ( int i =0; i < maxlen; i++){
            if(i < a.getDegree() && i < a.getDegree()){
                newCoefficients.add(a.coefficients.get(i) - b.getCoefficients().get(i));
            } else if(i > a.getDegree() && i < a.getDegree()){
                newCoefficients.add(-b.getCoefficients().get(i));
            } else if (i < actualsize && i > a.getDegree()){
                newCoefficients.add(a.coefficients.get(i));
            }
        }
        Polynomial result = new Polynomial(newCoefficients);
        return result;
    }
    public Polynomial addZeros(Polynomial a, int n){
        List<Integer> newCoefficients = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            newCoefficients.add(0);
        }
        for (int i = 0; i < a.getDegree() + 1; i++) {
            newCoefficients.add(a.coefficients.get(i));
        }
//        this.coefficients = newCoefficients;
        return new Polynomial(newCoefficients);
    }
}

File: mpi2/Polynomial.cpp
#include "Polynomial.h"
#include <vector>
#include <string>
#include <iostream>
#include <algorithm>
using namespace std;
class Polynomial {
public :

	vector<int> coefficients;
	int degree;

	Polynomial(int degree) {
		this->coefficients = vector<int>(degree, 0);
		this->degree = degree;
	}
	int getDegree() {
		return this->degree;
	}

	vector<int> getCoefficients() {
		return this->coefficients;
	}
	void setCoefficient(int index, int value) {
		this->coefficients[index] = value;
	}

	string toString() {
		string toOutput = "Polynomial with degree : ";
		toOutput +=  to_string(this->degree);
		toOutput += " and coefficients: ";
		for (auto it : this->coefficients)
		{
			toOutput += to_string(it);
			toOutput += ", ";
		}

		return toOutput;
	}
	void add(Polynomial other) {
		for (int i = 0; i < this->getCoefficients().size() - 1; i++) {
			this->coefficients[i] += other.getCoefficients()[i];
		}
	}
	Polynomial(vector<int> co) {
		this->coefficients = co;
		this->degree = co.size() - 1;
	}
};
/*
ostream& operator<<(ostream& Str, const Polynomial& v) {
	string toOutput = "Polynomial with degree : ";
	toOutput += v.degree;
	toOutput += " and coefficients: ";
	for (auto it : v.coefficients)
	{
		toOutput += it;
		toOutput += ", ";
	}

	return Str << toOutput;
}*/